### PR TITLE
fix(normalize): Allow override context.campaign

### LINKED ISSUE #34: Setting custom `campaign` properties
Hello.  I cannot seem to get custom campaign properties sent in `context` for a `page` call.

Here is the code on the page:

```js
!function(){var analytics=window.analytics=window.analytics||[];if(!analytics.initialize)if(analytics.invoked)window.console&&console.error&&console.error("Segment snippet included twice.");else{analytics.invoked=!0;analytics.methods=["trackSubmit","trackClick","trackLink","trackForm","pageview","identify","reset","group","track","ready","alias","debug","page","once","off","on"];analytics.factory=function(t){return function(){var e=Array.prototype.slice.call(arguments);e.unshift(t);analytics.push(e);return analytics}};for(var t=0;t<analytics.methods.length;t++){var e=analytics.methods[t];analytics[e]=analytics.factory(e)}analytics.load=function(t){var e=document.createElement("script");e.type="text/javascript";e.async=!0;e.src=("https:"===document.location.protocol?"https://":"http://")+"cdn.segment.com/analytics.js/v1/"+t+"/analytics.min.js";var n=document.getElementsByTagName("script")[0];n.parentNode.insertBefore(e,n)};analytics.SNIPPET_VERSION="4.0.0";
	analytics.load("<key>");

	function getQueryVariable(variable) {
    	var query = window.location.search.substring(1);
    	var vars = query.split("&");
    	for (var i=0;i<vars.length;i++) {
        	var pair = vars[i].split("=");
        	if(pair[0] == variable){return pair[1];}
    	}
    	return(void 0);
	}

	var campaign = {
        name: getQueryVariable("ctx_campaign"),
        source: getQueryVariable("ctx_source"),
        medium: getQueryVariable("ctx_medium"),
        term: getQueryVariable("ctx_term"),
        content: getQueryVariable("ctx_content")
    }

	window.console && window.console.log && window.console.log('campaign', campaign)

	analytics.page({}, {
      	context: {campaign: campaign},
     	campaign: campaign
	});
}}();
```

The campaign object prints correctly to the console. However, the API call does not include `campaign`:

```
{
  "context": {
    "campaign": {
      
    },
    "page": {
      "path": "\/offer",
      "referrer": "",
      "search": "?ctx_source=spaintrain&ctx_term=spain,train&ctx_campaign=spainzy",
      "title": "[Pricing - Step 1] Pricing.Full.Length.1.20",
      "url": "http:\/\/pricing-analysis.myhomezen.com\/offer?ctx_source=spaintrain&ctx_term=spain,train&ctx_campaign=spainzy"
    },
    "userAgent": "Mozilla\/5.0 (X11; Linux x86_64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/56.0.2924.87 Safari\/537.36",
    "library": {
      "name": "analytics.js",
      "version": "3.0.0"
    }
  },
  "integrations": {
    
  },
  "properties": {
    "path": "\/offer",
    "referrer": "",
    "search": "?ctx_source=spaintrain&ctx_term=spain,train&ctx_campaign=spainzy",
    "title": "[Pricing - Step 1] Pricing.Full.Length.1.20",
    "url": "http:\/\/pricing-analysis.myhomezen.com\/offer?ctx_source=spaintrain&ctx_term=spain,train&ctx_campaign=spainzy"
  },
  "category": null,
  "name": null,
  "anonymousId": "be16ec17-7f10-4f70-9072-df5491ce78ec",
  "timestamp": "2017-02-24T23:25:04.878Z",
  "type": "page",
  "writeKey": "<key>",
  "userId": "<id>",
  "sentAt": "2017-02-24T23:25:04.881Z",
  "_metadata": {
    "bundled": [
      "AdRoll",
      "Amplitude",
      "Facebook Pixel",
      "Google Analytics",
      "Google Tag Manager",
      "Lucky Orange",
      "Mixpanel",
      "Segment.io"
    ],
    "unbundled": [
      
    ]
  },
  "messageId": "ajs-55423f6e71370b8998553bd7d25c2009"
}
```

It is possible to see this live at https://pricing-analysis.myhomezen.com/offer?ctx_source=spaintrain&ctx_term=spain,train&ctx_campaign=spainzy

It seems to be supported, because I updated the page event test like so and it still passes:

```js
    it('should accept top level option .context', function() {
      var app = { name: 'segment' };
      var campaign = { name: 'segment' };
      analytics.page({ prop: true }, { context: { app: app, campaign: campaign } });
      var page = analytics._invoke.args[0][1];
      assert.deepEqual(app, page.obj.context.app);
      assert.deepEqual(campaign, page.obj.context.campaign);
    });
```

Is this intended to be supported?  If so, pointers on where to look next to find the issue?

### PR DESCRIPTION
**Pull request by SpainTrain**
- If an override context.campaign is provided, do not overwrite with UTM
  from querystring

closes https://github.com/segmentio/analytics.js-core/issues/34

### CODE REVIEW

**Comment by f2prateek** [comment]
LGTM. cc @stephenmathieson https://github.com/segmentio/analytics.js-core/issues/34#issuecomment-284589145

**Review by f2prateek — APPROVED** [review]

File: lib/index.js [hunk 1]
```diff
@@ -213,7 +213,9 @@ Segment.prototype.normalize = function(msg) {
       ctx.traits.crossDomainId = crossDomainId;
     }
   }
-  if (query) ctx.campaign = utm(query);
+  if (query && !ctx.campaign) {
+    ctx.campaign = utm(query);
```
Comment (f2prateek) [inline lib/index.js]: A possible improvement might be to do a merge if `ctx.campaign` exists. e.g. maybe you track referrer yourself, but still want the other context.campaign fields to be filled by us.

Comment (SpainTrain) [reply lib/index.js]: I started to do that initially, but then thought that campaign properties usually go together as a group, so there is some risk of inadvertently merging campaign properties from two different groups (if overrides and UTM differ).

happy to do it either way and I can make the change quickly if desired.

**Comment by SpainTrain** [comment]
hmm, first CI run passed, then the second (fixed tpyo in commit message, no code change) failed due to failing assertions about cookies in before/after hooks.

**Comment by SpainTrain** [comment]
> hmm, first CI run passed, then the second (fixed tpyo in commit message, no code change) failed due to failing assertions about cookies in before/after hooks.

disregard, kicking it off again made it pass.

File: lib/index.js [hunk 2]
```diff
@@ -213,7 +213,9 @@ Segment.prototype.normalize = function(msg) {
       ctx.traits.crossDomainId = crossDomainId;
     }
   }
-  if (query) ctx.campaign = utm(query);
+  if (query && !ctx.campaign) {
```
Comment (stephenmathieson) [inline lib/index.js]: I'd like to see a comment here explaining why we're doing this, as it isn't clear without digging

File: lib/index.js (hunk 1 above)
Comment (f2prateek) [reply lib/index.js]: Let's keep it as is for now.

File: lib/index.js (hunk 2 above)
Comment (f2prateek) [reply lib/index.js]: We don't want to override campaign if a user is providing one. This is fairly standard in most of our client libraries, including this one, no?

e.g.

1. https://github.com/SpainTrain/analytics.js-integration-segmentio/blob/bff98739474c1ce92809601cc4f76bb4acc6fad5/lib/index.js#L210
2. https://github.com/SpainTrain/analytics.js-integration-segmentio/blob/bff98739474c1ce92809601cc4f76bb4acc6fad5/lib/index.js#L207

Comment (SpainTrain) [reply lib/index.js]: Added the comment, but can remove based on what you all decide

File: lib/index.js (hunk 1 above)
Comment (SpainTrain) [reply lib/index.js]: :+1:

File: lib/index.js (hunk 2 above)
Comment (f2prateek) [reply lib/index.js]: Thanks —  adding a comment certainly doesn't hurt, just wasn't sure if it was required! No need to remove it now :)

**Comment by SpainTrain** [comment]
@stephenmathieson anything else I should fix here? Thanks for the review!

**Review by stephenmathieson — APPROVED** [review]

**Comment by stephenmathieson** [comment]
Nope, LGTM!

**Comment by SpainTrain** [comment]
Awesome.

You guys good to merge and ship?

**Comment by stephenmathieson** [comment]
@f2prateek I'm not sure what the process for releasing changes is for this integration. Any way you could handle that?

**Comment by f2prateek** [comment]
@stephenmathieson Sure, for future reference it is https://paper.dropbox.com/doc/Analytics.js-Deployment-ZFzuZSXuTQC.

**Comment by stephenmathieson** [comment]
Ah ok, thanks dude!

**Comment by SpainTrain** [comment]
:tada:

**Comment by SpainTrain** [comment]
What is the expected latency before these changes will be in the production asset?  Not trying to be pushy, just want to know when I should test my custom `campaign` tracking code again.  Thanks!

**Comment by f2prateek** [comment]
@SpainTrain generally takes about a couple of hours, but I'm going to sneak in a few other changes that have been pending for a while as well though; should be no later than Monday (want to avoid deploying on Friday).

**Pull request merged**